### PR TITLE
Allow collection of EnsoGL stats when Monitor panel is not visible

### PR DESCRIPTION
**Pull request by akavel**
### Pull Request Description

This change makes EnsoGL runtime stats be always collected, even when EnsoGL `Monitor` panel is not visible. Those stats are intended to be used in the future by a profiling framework.

**Performance impact:** Continuous collection of stats introduces an overhead of two Web Performance API `now()` calls in each frame of the main rendering loop, plus a small number of simple arithmetic calculations. This is assumed to be a negligible and acceptable overhead.

#### Visuals

A screenshot of the Monitor panel in full `ide` after applying the PR, taken in IDE built with `./run dist`:

Screenshot 2022-02-14 at 16 11 42

A recording, also in IDE built with `./run dist`; note that FPS is impacted by the act of recording itself:

https://user-images.githubusercontent.com/273837/154104016-49a12e23-1210-4477-9743-ec1611e5b4ed.mov

https://www.pivotaltracker.com/story/show/181093601

### Important Notes

 - Responsibility for controlling how `Stats` gathering and calculation is performed at various points in the main rendering loop was removed from `Monitor` - the `Monitor`'s purpose is only to display existing data, it should not influence how the data is collected.
 - Two previously existing distinct `Monitor` structs were merged into one, to avoid confusion; after previous refactorings, the remaining `stats::Monitor` did not have much useful code anyway.
 - In `stats` package, refactoring was done, to make `StatsData` a "dumb", data-only type, and to move the logic related to stats collection and frame tracking to other helper types.

[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

**Review by wdanilo — CHANGES_REQUESTED** [review]
I don't understand this design and thus I don't like it - why this is implemented as a "switch" in the World struct? Who needs it? What is the desired usage scenario? Why the World struct is responsible for remembering it? Are we sure that this is the desired design? I believe these questions are **impossibe** to be answered without seeing a code that requires this functionality – maybe the responsibilities should belong to another entity?

This is another example of PR which implements dead code (similar to the previous one). I am not sure if this is the right design. I would like to see the usage of this functionality in the profiling framework first in order to understand if this should be implemented this way or other way around. I'm not sure whether this should be implemented as a "switch" in the World struct, or maybe profiling framework should manage it somehow else. And it's impossible to decide without seeing the code in context of its usage.

I would like all of our PRs to never be implemented as a dead code (like this PR is), always with a clear usage scenario. The design of the code should be dictated by how the thing is intended to be used. Right now, we have several super small changes with new structures and APIs not used by anyone and IMO there is a high probability that after the final PR will be created with the profiling framework, I will request changes to the API implemented in these PRs, because the API should be judged by the usage scenario. So I'd love ask to close this PR and implement PRs with code changes that are required by real required usage, not artificial debug scenes.

File: lib/rust/ensogl/core/src/display/world/stats.rs
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/display/world/stats.rs]: Should there not also be a switch to disable this?

**Comment by akavel** [comment]
@wdanilo I refactored the code as inspired by our discussions on chat (the video & screenshot are also updated); PTAL if it makes more sense now. Thanks for all the valuable lessons!

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (MichaelMauderer) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: Could this have the return type `&[Panel]`?

**Review by MichaelMauderer — APPROVED** [review]

File: lib/rust/ensogl/core/src/display/world.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/display/world.rs]: Is this world's responsibility to measure the time and provide the current time value to stats collector? Why world is handling that, but for example resetting the stats per frame is handled by stats collector?

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: This is a bad pattern to borrow_mut for the whole scope. Someone will refactor this code in the future, put her borrows in the lines below and it will blow up. Better way of doing it is the way that function `toggle` is implemented - borrow once and redirect the call to the borrowed structure.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: I was writing about it in another PR review – I do not understand why this field is here. Also, I think its name is very confusing. This is frame counter of only these frames where the stats were gathered, this is not "all frames counter", as the name may suggest.

Moreover, it seems that the only place this variable is being used is when it is compared to 0 to check if this is the first sample, so why we are collecting information of rendered frames if no one uses this information?

Comment (wdanilo) [reply lib/rust/ensogl/core/src/debug/stats.rs]: Also, no docs describe that stats collector assumes that it is running in every frame – that was explained on Discord, but this information is missing from the code.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/stats.rs]: **Re: `frame_counter`:** I changed the field to a boolean flag, thanks! Other than that, I'm not fully sure if what you were commenting on is just that it was a number, or that it exists at all? If the latter, then I don't currently see good alternatives:

 - I could add an `is_default()` method to `StatsData`, checking that every field equals zero, but that would be much more complex and also less effective;
 - I could add an `is_valid()` method checking e.g. just some fields (`self.fps != 0.0 || self.frame_counter != 0.0`), but I'm not convinced that would be more clear; an `initialized` field (or whatever name we'll agree upon) sounds most direct as to intention;
 - I can still name the field differently than `initialized`; however, e.g. `is_first_sample` doesn't sound appropriate to me, as this is notably not the *first* frame/sample that is "invalid", but the "0-th" - i.e., "the predecessor to the 1st frame" - and naming it `is_frame_zero` also sounds confusing to me, esp. given that programmers at various times count from 1 or from 0, so it's still not clear if 0 is "first" or not... other than that, I can't think of any other good names for it - I initially thought of `valid`, but I find it arguable which one is more clear.

Comment (kazcw) [reply lib/rust/ensogl/core/src/debug/stats.rs]: What about `Option<StatsData>`?

File: lib/rust/ensogl/core/src/display/world.rs
Comment (kazcw) [reply lib/rust/ensogl/core/src/display/world.rs]: I think this inconsistency is because `begin_frame` is misplaced. It is an operation for building up a sequence of `StatsData` values; it seems odd for sequence logic to be in a method on a `StatsData` object.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (akavel) [reply lib/rust/ensogl/core/src/debug/stats.rs]: ~~@kazcw Initially (some time ago) I suggested to return `Option<StatsData>` from `begin_frame()`, but Wojciech then suggested to rather return `default()` than `None` in case of the "uninitialized" frame. Would your suggestion be to track this as `Option<StatsData>` inside, but then return `default()` from `begin_frame()` instead of `None`? Or to return `None` from `begin_frame()` in that case?~~

I modified the code to return `Option<StatsData>` from `begin_frame()`, as I believe this is indeed a better design.

**Comment by akavel** [comment]
@wdanilo @kazcw PTAL; inspired by a chat with Kaz on Friday, I modified the code to what I believe is a better design.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: Errors should be valid English sentences. Please apply to other errors too.

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: What do you mean by "there is no frame before "Frame 1""? Before the first rendering frame the program can evaluate expressions, so there is time before frame 1.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: This should be refactored to other abstractions over web-functions (I dont remember the module where we keep them).

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: this function is huge. The previous design with `reset_per_frame_statistics` was nicer because it made such functions easier to read.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: the var naming is strange here. This var does not contain a new frame, it's misleading.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: this can be implemented as a mapping for Option, which we always prefer instead of writing `None => None`.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: I'm not sure whether this struct is collecting stats. It has a reference to StatsData (which contains stats) and then it cleans the stats on every frame, and adds info about memory usage and frame time after the frame. I would not name it a "stats collector" in such a case.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: The wording "are called properly" is not explained. What does "properly" mean?

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: I don't understand why a "stats collector" which is put in Rc<RefCell<...>> is named `StatsOverTime`. So a shared stats collector are stats over time?

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: What does mean "stats related to the current frame"?

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs:846]: This is a nice design!

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/monitor.rs]: please extract all constants, like this `0.001` to well named variables explaining what they are.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (wdanilo) [inline lib/rust/ensogl/core/src/debug/stats.rs]: Please do not use constraints in structs definition. I was explaining it multiple times in different PRs. Such constraints leak everywhere. For example, you implement later `Clone` for this structure. Is it really necessary that `T:TimeProvider` when cloning the structure?

There are big topics about it on Haskell mailing lists and it is considered a big anti-pattern in Haskell (which is supposed to be banned). I believe that we should ban it too.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (MichaelMauderer) [reply lib/rust/ensogl/core/src/debug/stats.rs]: Alternatives `FrameStatsData`.

File: lib/rust/ensogl/core/src/debug/stats.rs
Comment (MichaelMauderer) [reply lib/rust/ensogl/core/src/debug/stats.rs]: Alternatives `SharedStatsHandler`. (Or `Shared`...whatever the inner struct is).

File: lib/rust/ensogl/core/src/debug/monitor.rs
Comment (MichaelMauderer) [reply lib/rust/ensogl/core/src/debug/monitor.rs]: Needs reworking of the comment and/or additional check for `None` value.

**Review by wdanilo — APPROVED** [review]

**Comment by akavel** [comment]
(*Process note:* the code-review is now accepted, therefore the PR is waiting for Review by @MichaelMauderer on Task https://www.pivotaltracker.com/story/show/181093601 before merge.)

**Pull request merged**
